src/compute/src/compute_state.rs

```diff
@@ -146,11 +155,10 @@ impl ComputeState {
             command_history,
             max_result_size: u64::MAX,
             dataflow_max_inflight_bytes: None,
-            linear_join_spec: Default::default(),
             metrics,
             tracing_handle,
-            enable_operator_hydration_status_logging: true,
             context,
+            worker_config: mz_dyncfgs::all_dyncfgs(),
```

Comment:
If we do end up moving the compute dyncfg defs out of mz_dyncfgs and into mz_compute_types, then this `all_dyncfgs()` probably wants to become the compute ones + the persist ones

Reply:
Can you elaborate on this?

AFAICT the set of configs in `worker_config` must always match the set of configs sent by envd. Otherwise you get errors like this one:
```
ERROR mz_dyncfg: config update persist_txns_data_shard_retryer_clamp Some(Duration(ProtoDuration { secs: 16, nanos: 0 })) not known set: {"enable_columnation_lgalloc": Bool(false), "enable_compute_chunked_stack": Bool(false), "enable_compute_operator_hydration_status_logging": Bool(true), "enable_lgalloc_eager_reclamation": Bool(true), "enable_mz_join_core": Bool(true), "linear_join_yielding":String(RwLock { data: "work:1000000,time:100", poisoned: false, .. })}
```

I assume these come from attempting to apply `ConfigUpdates` to a `ConfigSet` that doesn't know all the contained configs.

Currently envd sends all dyncfgs it knows about to the clusterds. Is your suggestion to change this to send only the configs clusterd needs? Or to have a way to apply `ConfigUpdates` in a way that ignores unknown configs?

That latter would let us reduce `worker_config` to `mz_compute_types::dyncfgs::all_dyncfgs()` only. There is actually no reason to have the persist configs in that set too.

Reply:
Ah yeah, sorry, I blame jetlag-brain. Forget what I said above.

One option (I think my preferred one) is to seed this with something that has the full set of dyncfgs registered. We want to make sure they start with the same value across all workers in a replica, which means we have to care about process boundaries. That probably means we want to construct a set of every config and its default value up in clusterd's `fn main()` and plumb it down (relying on every process in a replica running the same version of code and so having the same compiled in default). That means that what you had before (`mz_dyncfgs::all_dyncfgs`) is exactly correct, but instead called higher up in the dep tree than you had and passed down. Hmm, slightly awkward, but we only have to get it right once?

> Or to have a way to apply ConfigUpdates in a way that ignores unknown configs?

This would also work, but it's pretty easy to accidentally forget to add a new cfg to the all_dyn_configs list (speaking from experience). If we did this option, we'd lose the WARN and it would silently never update, which would be pretty bad.

Reply:
> We want to make sure they start with the same value across all workers in a replica

Doesn't making each worker call `mz_dyncfgs::all_dyncfgs()` do that already? Each of the per-worker `ConfigSet`s will be initialized with the defaults hardcoded in the `Config` definitions. What could go wrong?

> That probably means we want to construct a set of every config and its default value up in clusterd's `fn main()` and plumb it down

Is your suggestion that `main` should create a single `ConfigSet` and pass that to all workers:

```rust
let config = mz_dyncfgs::all_dyncfgs();
for worker in workers {
    spawn_worker(worker, config.clone());
}
```

... or that `main` creates a separate `ConfigSet` for each worker:

```rust
for worker in workers {
    spawn_worker(worker, mz_dyncfgs::all_dyncfgs());
}
```

I think the former has the defect that all workers share the same config values, so updates would propagate to every worker immediately. And the latter is the same as we have now, just with the `ConfigSet`s created in a different place.

Reply:
Ah, sorry, I think I was just being unclear. What you merged here is exactly what I was saying, with the small exception that I don't think mz_compute->mz_dyncfgs is a great dep. Ideally mz_dyncfgs is only depended on at the very leaf packages (i.e. the `fn main`s) and then the result of calling it passed down through constructors.

I might just type this change up myself in a PR and also include a fix for the N^2 Moritz pointed out as well as doc updates to adjust the singleton language. It'll likely be a couple days, but definitely fine to leave things as-is even it if takes longer than that.

Thanks again for hooking this up in compute!!

Reply:
Ah, got it!

Thanks for doing all the actual hard work! This is so much nicer than what we had before :)